docs/design-proposals/2023-11 TraceQL Metrics.md

```diff
@@ -0,0 +1,112 @@
+  - [Step Interval](#step-interval)
+
+## Intro
+
+Let's look at an example straight away.  This query plots the error rate by service over time:  
+`{status = error } | rate() by (resource.service.name)`
+
+A normal TraceQL query finds matching spans and returns data points like time, duration, service name, status code, etc.  A metrics query at the core does the same thing: _find matching spans_, but instead of seeing individual results, they are aggregated over time to create time-series.  A metrics query is built by extending a regular query with metrics pipeline stages.  The structure looks like this:
+
+`<span query...> | <metrics query...>`
+
+An important understanding is that _any_ valid TraceQL query can be turned into a metrics query. This finds the call rates to service B but only when initiated upstream by service A:  
+`{resource.service.name="A"} >> {resouce.service.name="B"} | rate()`
+
+
+## First Stage
+
+The first stage in a metrics query is responsible for turning spans into time series. This is a unique intersection between the two data types and has specialized aggregation functions.  We use the `..._over_time()` suffix to distinguish them from their counterparts.
+
+| function                                   | description                |
+---------------------------------------------| ---------------------------- 
+|`rate()`                                    | The rate of spans per second
+|`count_over_time()`                         | The count of spans
+|`avg_over_time(<field>)`                    | The average of a numeric value like `duration` or [http.request.body.size](https://opentelemetry.io/docs/specs/semconv/attributes-registry/http/)
+|`max_over_time(<field>)`                    | The maximum of a numeric value.
+|`min_over_time(<field>)`                    | The minimum of a numeric value.
+|`quantile_over_time(<field>, q1, q2, ...)`  | The quantile of a numeric value (e.g. p95). Multiple quantiles can be requested and a time-series is created for each. |
+
+### Interval
+All aggregations take an optional interval like `30s`, `5m`, `1h`. If not specified then the interval will automatically match the step interval of the query. The interval is the last parameter of each function.
+
+`rate(5m)`  
+`quantile_over_time(duration, 0.95, 5m)`
+
+### Grouping
+All aggregations can group by one or more attributes and generate a time-series for each combination of values.  
+
+Plot the request rate of `/myapi` by user ID and HTTP status code:  
+`{ span.http.path = "/myapi" } | rate() by (span.user_id, span.http.status_code)`
+
+Plot the p95 duration of requests to service by HTTP endpoint:  
+`{ resource.service.name = "myservice" } | quantile_over_time(duration, 0.95) by (span.http.path)`
+
+### Filtering
+Aggregations can be combined with comparison operations to filter the output.
+
+Find the request rate per service but only if more than 1000 req/s:  
+`{ } | rate() by (resource.service.name) > 1000`
```

Comment:
Syntax wise this is a bit confusing for me. Not sure how to make it better but that is less confusing for me to read

```
({ } | rate() by (resource.service.name)) > 1000
```

Becomes even weirder when you have 

```
{ } | rate() by (resource.service.name) > {} |  rate() by (resource.bar)
```

Reply:
In the first case the comparison operator is part of the pipeline stage: `{ } | (rate() by (resource.service.name) > 1000)` so you can drop data but keep going.  `{} | rate() > 1000 | max() > 2000`.  I think this is equivalent to the functional style `max(rate({}) > 1000) > 2000)`.

Cross-series query:  agree that syntax could be clearer.  Although in this case I'm not sure that query works since they have different labels.  We could add `on(x,y,z)` to specify the join conditions like in PromQL.  But in the first release we simplify and require same labels.

Reply:
I feel like the second example only returns one series with `resource.bar` having no value. The attribute `resource.bar` has already been aggregated away in the stage before it.

Actually, is the second example even allowed? `rate()` should require an input of a spanset pipeline. I don't think it can take metrics.

Reply:
@joe-elliott  The second example is a comparison between series:  `>`.  Easy to miss? Which is further confirmation the syntax would be better.   You are correct if it was a normal multi-stage pipeline.   We don't plan to "backfill" the `by` clauses so `by(a) | by(b)` doesn't work you need `by(a,b) | by(b)`

Reply:
ah, I see! it is easy to miss. i'll point out that spanset pipelines currently require parentheses to be parsed:

https://github.com/grafana/tempo/blob/5e7023521e080153e7af83017c7cc9a0acd963c8/pkg/traceql/expr.y#L120-L124

This is not valid:
```
{ span.http.status_code = 200 } | count() > { span.http.status_code = 500 } | count()
```
It must be this:
```
({ span.http.status_code = 200 } | count()) > ({ span.http.status_code = 500 } | count())
```

I assume metrics would follow suit? iirc the `()` requirement was added to prevent ambiguity in the parser.

Reply:
Yes this is likely but won't be certain until we try to implement the parsing changes.  Do you think the proposal should be updated to reflect this now?